### PR TITLE
Implement support for Null values in columns

### PR DESCRIPTION
**Pull request by pgovind**
Implements support for null values in columns. 
Updated the unit tests to include null values and added some new ones to test the more corner cases

### CODE REVIEW

File: src/Microsoft.Data/BaseColumn.cs [hunk 1]
```diff
@@ -64,17 +68,19 @@ protected set
...
 
+        internal virtual BaseColumn CloneAndAppendNulls(BaseColumn mapIndices = null, bool invertMapIndices = false) => throw new NotImplementedException();
+
         public virtual BaseColumn Sort(bool ascending = true) => throw new NotImplementedException();
 
         internal virtual BaseColumn GetAscendingSortIndices() => throw new NotImplementedException();
 
         internal delegate long GetBufferSortIndex(int bufferIndex, int sortIndex);
-        internal delegate T GetValueAtBuffer<T>(int bufferIndex, int valueIndex);
+        internal delegate Tuple<T, int> GetValueAndBufferSortIndexAtBuffer<T>(int bufferIndex, int valueIndex);
```
Comment (gfoidl) [inline src/Microsoft.Data/BaseColumn.cs]: Can this be a value-tuple? Would safe the allocation for `Tuple<T1, T2>`.

File: src/Microsoft.Data/PrimitiveColumn.Sort.cs [hunk 2]
```diff
@@ -36,26 +37,43 @@ private void GetSortIndices(IComparer<T> comparer, out PrimitiveColumn<long> col
                 bufferSortIndices.Add(sortIndices);
             }
             // Simple merge sort to build the full column's sort indices
+            Tuple<T, int> GetFirstNonNullValueAndBufferIndexStartingAtIndex(int bufferIndex, int startIndex)
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumn.Sort.cs]: Value-tuple?

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 3]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+            lastBuffer.Append(value ?? default);
+            SetBit(Length, value.HasValue ? true : false);
             Length++;
         }
 
+        public bool IsValid(long index) => NullCount == 0 || GetBit(index);
+
+        private void SetBit(long index, bool value)
+        {
+            if (index < 0 || index > Length)
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]:
```suggestion
            if ((ulong)index > Length)
```
Saves a explicit comparison. Negative indexed casted to `ulong` results in values > long.MaxValue.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 4]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]: Maybe use a ThrowHelper? Altough this method is quite long, so won't be inlined. Still the JITed code will be smaller, as the creation of the exception is moved out of this method.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 5]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+            }
+            // First find the right bitMapBuffer
+            int bitMapIndex = (int)(index / Buffers[0].MaxCapacity);
+            Debug.Assert(NullBitMapBuffers.Count > bitMapIndex);
+            DataFrameBuffer<byte> bitMapBuffer = NullBitMapBuffers[bitMapIndex];
+
+            // Set the bit
+            index -= bitMapIndex * Buffers[0].MaxCapacity;
+            int bitMapBufferIndex = (int)index / 8;
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]:
```suggestion
            int bitMapBufferIndex = (int)index >> 3;    // index / 8
```
The bit-shift will produce [better code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAbgFgAoYgZkoCZyBhcgb0fIGVWASwB2GcgFkAFABsIogObkxAExgIAlOQC8APnLSxGTWo3kA9OQAc9BoPL9BLFePIA5OQuVmtug0bipqLqCOR6Bsx2AL5AA=)

JIT will substitute division by a power of two only for unsigned ints / longs. So instead of the bit-shift it [could also be](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAbgFgAoYgZkoCZyBhcgb0fIGVWASwB2GcgFkAFABsIogObkxAExgIAlOQC8APnLSxGTWo3kA9OQAc9BoPL9BLFePIA5OQuVmtug0bipqLqCOR6Bsx2Dk4CLsZSVF5KrqHa+obGmtLSAK5ZvpY2mnYAvkA===)
```suggestion
            int bitMapBufferIndex = (int)((uint)index / 8);
```

You can take whatever you prefer :wink:
I would take the latter.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 6]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+            Debug.Assert(bitMapBuffer.Length >= bitMapBufferIndex);
+            if (bitMapBuffer.Length == bitMapBufferIndex)
+                bitMapBuffer.Append(0);
+            byte curBitMap = bitMapBuffer[bitMapBufferIndex];
+            byte newBitMap;
+            if (value)
+            {
+                newBitMap = (byte)(curBitMap | (byte)(1 << (int)(index % 8)));
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]:
```suggestion
                newBitMap = (byte)(curBitMap | (byte)(1 << (int)((uint)index % 8)));
```
Boils down to [index & 7](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAbgFgAoYgZkoCZyBhcgb0fIGVWASwB2GcgFkAFABsIogObkxAExgIAlOQC8APnLSxGTUdHqE5AKTkAHJvoNB5foJYrx5AHJyFytRra+obGptIArqEBljb2jgC+QA).

Or you write the bit-hack yourself as `(int)(index & 7)`.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 7]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+                if ((curBitMap >> ((int)(index % 8)) & 1) == 0 && index < Length && NullCount > 0)
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]:
```suggestion
                if ((curBitMap >> ((int)index & 1) == 0 && index < Length && NullCount > 0)
```
`index % 8` <=> `index & 7`, so `& 1` will mask the rest, just `& 1` remains.

JIT [won't optimize this](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAbgFgAoYgZkoCZyBhcgb0fIGVWASwB2GcgFkAFABsIogObkxAExgIAlOQC8APnLTpYjJuOj1CcgFJyADk3aAZOVr9B7gSxXjyAOTkFZTUNbX1Dc1MjAFcTTRCrWwdnV3oGQXJPIR8JSSpApR9LMINIswTyFwB2FLcGAF8gA=).

Same below in else-branch.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 8]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+                { 
+                    // old value was NOT null and new value is null
+                    NullCount++;
+                }
+                else if (index == Length)
+                {
+                    // New entry from an append
+                    NullCount++;
+                }
+                newBitMap = (byte)(curBitMap & (byte)~(1 << (int)(index % 8)));
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]:
```suggestion
                newBitMap = (byte)(curBitMap & (byte)~(1 << (int)((uint)index % 8)));
```

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 9]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+            }
+            bitMapBuffer[bitMapBufferIndex] = newBitMap;
+        }
+
+        private bool GetBit(long index)
+        {
+            if (index < 0 || index > Length)
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]:
```suggestion
            if ((ulong)index > Length)
```
See above. Also for ThrowHelper-thoughts.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 10]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            // First find the right bitMapBuffer
+            int bitMapIndex = (int)(index / Buffers[0].MaxCapacity);
+            Debug.Assert(NullBitMapBuffers.Count > bitMapIndex);
+            DataFrameBuffer<byte> bitMapBuffer = NullBitMapBuffers[bitMapIndex];
+
+            // Get the bit
+            index -= bitMapIndex * Buffers[0].MaxCapacity;
+            int bitMapBufferIndex = (int)index / 8;
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]:
```suggestion
            int bitMapBufferIndex = (int)(uint)index / 8;
```

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 11]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+            Debug.Assert(bitMapBuffer.Length > bitMapBufferIndex);
+            byte curBitMap = bitMapBuffer[bitMapBufferIndex];
+            return ((curBitMap >> ((int)index % 8)) & 1) != 0;
```
Comment (gfoidl) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]:
```suggestion
            return ((curBitMap >> ((int)index & 1) != 0;
```
See above.

File: src/Microsoft.Data/StringColumn.cs [hunk 12]
```diff
@@ -124,47 +138,64 @@ protected override void SetValue(long rowIndex, object value)
...
+                {
                     sortIndices[i] = i;
+                }
                 // TODO: Refactor the sort routine to also work with IList?
                 string[] array = buffer.ToArray();
                 IntrospectiveSort(array, array.Length, sortIndices, comparer);
                 bufferSortIndices.Add(sortIndices);
             }
             // Simple merge sort to build the full column's sort indices
+            Tuple<string, int> GetFirstNonNullValueStartingAtIndex(int stringBufferIndex, int startIndex)
```
Comment (gfoidl) [inline src/Microsoft.Data/StringColumn.cs]: Valuetuple?

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs [hunk 13]
```diff
@@ -15,6 +15,8 @@ public static DataFrame MakeDataFrameWithTwoColumns(int length)
         {
             BaseColumn dataFrameColumn1 = new PrimitiveColumn<int>("Int1", Enumerable.Range(0, length).Select(x => x));
             BaseColumn dataFrameColumn2 = new PrimitiveColumn<int>("Int2", Enumerable.Range(10, length).Select(x => x));
+            dataFrameColumn1[length/2] = null;
```
Comment (gfoidl) [inline tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: Nit: spacing

```suggestion
            dataFrameColumn1[length / 2] = null;
```
Same on next line.

File: src/Microsoft.Data/BaseColumn.cs [hunk 14]
```diff
@@ -34,7 +34,11 @@ protected set
             }
         }
 
-        public long NullCount { get; protected set; }
+        public virtual long NullCount
```
Comment (eerhardt) [inline src/Microsoft.Data/BaseColumn.cs]: Why not `abstract`?  I guess the same could be said for GetValue and SetValue below.  (Should have said that last time)

File: src/Microsoft.Data/DataFrame.cs [hunk 15]
```diff
@@ -131,7 +132,9 @@ public DataFrame Sort(string columnName, bool ascending = true)
             List<BaseColumn> newColumns = new List<BaseColumn>(ColumnCount);
             for (int i = 0; i < ColumnCount; i++)
             {
-                var newColumn = Column(i).Clone(sortIndices, !ascending);
+                BaseColumn oldColumn = Column(i);
+                BaseColumn newColumn = oldColumn.CloneAndAppendNulls(sortIndices, !ascending);
```
Comment (eerhardt) [inline src/Microsoft.Data/DataFrame.cs]: `CloneAndAppendNulls` is internal. How could someone derive their own column type from BaseColumn, and then use this Sort method?

File: src/Microsoft.Data/BaseColumn.cs [hunk 16]
```diff
@@ -34,7 +34,11 @@ protected set
...
+        {
+            get => throw new NotImplementedException();
+            protected set => throw new NotImplementedException();
```
Comment (eerhardt) [inline src/Microsoft.Data/BaseColumn.cs]: Who would ever `set` this property? Should it be `get` only?

File: src/Microsoft.Data/PrimitiveColumn.cs [hunk 17]
```diff
@@ -106,13 +122,23 @@ public override BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapInd
             {
                 if (mapIndices.DataType != typeof(long))
                     throw new ArgumentException(Strings.MismatchedValueType + " PrimitiveColumn<long>", nameof(mapIndices));
-                if (mapIndices.Length != Length)
+                if (mapIndices.Length >= Length)
                     throw new ArgumentException(Strings.MismatchedColumnLengths, nameof(mapIndices));
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveColumn.cs]: Does `MismatchedColumnLengths` provide a good error message here now? Before it was just different lengths, but now the check is only for greater than or equals.

Same applies to the below usages of MismatchedColumnLengths.

File: src/Microsoft.Data/PrimitiveColumn.cs [hunk 18]
```diff
@@ -106,13 +122,23 @@ public override BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapInd
...
             }
             return Clone();
         }
 
+        internal override BaseColumn CloneAndAppendNulls(BaseColumn mapIndices = null, bool invertMapIndices = false)
+        {
+            PrimitiveColumn<T> ret = Clone(mapIndices, invertMapIndices) as PrimitiveColumn<T>;
+            for (long i = 0; i < NullCount; i++)
+            {
+                ret.Append(null);
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveColumn.cs]: Instead of appending `null` multiple times, what do you think about a method that will do it all at once?  Like `AppendMany(value, count)`.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 19]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
             }
-            lastBuffer.Append(value);
+            lastBuffer.Append(value ?? default);
+            SetBit(Length, value.HasValue ? true : false);
             Length++;
         }
 
+        public bool IsValid(long index) => NullCount == 0 || GetBit(index);
+
+        private void SetBit(long index, bool value)
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]: Is there a better name than `SetBit` for this method? I don't think it is very descriptive of what it does.  maybe `SetNullIndex`? or something linking it to `Null` support.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 20]
```diff
@@ -25,6 +31,7 @@ public PrimitiveColumnContainer(T[] values)
             {
                 curBuffer = new DataFrameBuffer<T>();
                 Buffers.Add(curBuffer);
+                NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]: Is it possible to lazy initialize the NullBitMapBuffers? For example, if I don't have any null values, there should be no needed to create these objects.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 21]
```diff
@@ -36,8 +43,10 @@ public PrimitiveColumnContainer(T[] values)
                 {
                     curBuffer = new DataFrameBuffer<T>();
                     Buffers.Add(curBuffer);
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
                 }
                 curBuffer.Append(values[i]);
+                SetBit(Length, true);
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]: Why are we doing this? It shouldn't be possible to get or set the null value at index `Length`. That should be out of range.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 22]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
+            // Get the bit
+            index -= bitMapIndex * Buffers[0].MaxCapacity;
+            int bitMapBufferIndex = (int)index / 8;
+            Debug.Assert(bitMapBuffer.Length > bitMapBufferIndex);
+            byte curBitMap = bitMapBuffer[bitMapBufferIndex];
+            return ((curBitMap >> ((int)index % 8)) & 1) != 0;
+        }
+
         public long Length;
         //TODO:
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]: Can you remove the TODO?

File: src/Microsoft.Data/PrimitiveColumnContainer.cs [hunk 23]
```diff
@@ -69,91 +81,180 @@ public PrimitiveColumnContainer(long length = 0)
...
-                int arrayIndex = GetArrayContainingRowIndex(ref startIndex);
-                bool temp = Buffers[arrayIndex][(int)startIndex, length, ret];
-                while (ret.Count < length)
+                long endIndex = Math.Min(Length, startIndex + length);
+                for (long i = startIndex; i < endIndex; i++)
                 {
-                    arrayIndex++;
-                    temp = Buffers[arrayIndex][0, length - ret.Count, ret];
+                    T? value = this[i];
+                    if (value.HasValue)
```
Comment (eerhardt) [inline src/Microsoft.Data/PrimitiveColumnContainer.cs]: Is this correct - to skip any nulls? Would the caller expect that passing in a `length` of `10` be returned with a list of `8` values?

File: src/Microsoft.Data/StringColumn.cs [hunk 24]
```diff
@@ -18,7 +16,17 @@ public partial class StringColumn : BaseColumn
...
-            _stringBuffers.Add(new List<string>());
+            int numberOfBuffersRequired = Math.Max((int)(length / int.MaxValue), 1);
+            for (int i = 0; i < numberOfBuffersRequired; i++)
+            {
+                long bufferLen = length - _stringBuffers.Count * int.MaxValue;
+                List<string> buffer = new List<string>((int)Math.Min(int.MaxValue, bufferLen));
+                _stringBuffers.Add(buffer);
+                for (int j = 0; j < bufferLen; j++)
+                {
+                    buffer.Add(default);
```
Comment (eerhardt) [inline src/Microsoft.Data/StringColumn.cs]: Is this necessary? If yes - is this the most performant way of appending `bufferLen` `nulls` to the list?

File: src/Microsoft.Data/StringColumn.cs [hunk 25]
```diff
@@ -85,10 +91,18 @@ protected override object GetValue(long startIndex, int length)
...
+            if (value == null || value is string)
             {
                 int bufferIndex = GetBufferIndexContainingRowIndex(ref rowIndex);
+                var oldValue = this[rowIndex];
                 _stringBuffers[bufferIndex][(int)rowIndex] = (string)value;
+                if (oldValue != (string)value)
+                {
+                    if (value == null)
+                        _nullCount++;
+                    if (oldValue == null && _nullCount > 0)
```
Comment (eerhardt) [inline src/Microsoft.Data/StringColumn.cs]: Can the `oldValue` be `null` AND `_nullCount <= 0`? Maybe that should be an `Assert`?

**Comment by eerhardt** [comment]
> using System;

copyright

---
Refers to: tests/Microsoft.Data.DataFrame.Tests/BufferTests.cs:1 in 9ce1ada. [](commit_id = 9ce1ada5149427c535e2d8b80aab01784efa747c, deletion_comment = False)

**Comment by eerhardt** [comment]
Maybe a good test to have would be that you can take an `Apache.Arrow.RecordBatch` and wrap a `DataFrame` around it without any copies. I'm wondering if the "null" support you are adding is compatible with the Arrow "null" format.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs (hunk 7 above)
Comment (pgovind) [reply src/Microsoft.Data/PrimitiveColumnContainer.cs]: The `& 1` should take place after the ` >> ` :)

File: src/Microsoft.Data/PrimitiveColumnContainer.cs (hunk 11 above)
Comment (pgovind) [reply src/Microsoft.Data/PrimitiveColumnContainer.cs]: Same reply. The right shift should happen before the And

File: tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs (hunk 13 above)
Comment (pgovind) [reply tests/Microsoft.Data.DataFrame.Tests/DataFrameTests.cs]: Damn, I wonder why the .editorconfig did not find this

File: src/Microsoft.Data/PrimitiveColumnContainer.cs (hunk 7 above)
Comment (gfoidl) [reply src/Microsoft.Data/PrimitiveColumnContainer.cs]: Ups... :smile:

File: src/Microsoft.Data/DataFrame.cs (hunk 15 above)
Comment (pgovind) [reply src/Microsoft.Data/DataFrame.cs]: Good catch!

File: src/Microsoft.Data/PrimitiveColumn.cs (hunk 18 above)
Comment (pgovind) [reply src/Microsoft.Data/PrimitiveColumn.cs]: Yup, good idea. Done

File: src/Microsoft.Data/PrimitiveColumnContainer.cs (hunk 19 above)
Comment (pgovind) [reply src/Microsoft.Data/PrimitiveColumnContainer.cs]: How about `SetValidityBit`? Unfortunately in Arrow, a `1` implies a valid value and a `0` implies a null value.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs (hunk 20 above)
Comment (pgovind) [reply src/Microsoft.Data/PrimitiveColumnContainer.cs]: I did consider lazy initialization, but the C# Arrow implementation seems to always maintain a null bit map, so I went with this.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs (hunk 21 above)
Comment (pgovind) [reply src/Microsoft.Data/PrimitiveColumnContainer.cs]: Right, `GetValidityBit` does throw an exception when `index >= Length`. The set here is still valid because the current length is `Length + 1`, after the `curBuffer.Append` call just above. Line 50(the next line) is `Length++` and updates the length.

File: src/Microsoft.Data/PrimitiveColumnContainer.cs (hunk 23 above)
Comment (pgovind) [reply src/Microsoft.Data/PrimitiveColumnContainer.cs]: Yup, this was wrong. Fixed it

File: src/Microsoft.Data/StringColumn.cs (hunk 24 above)
Comment (pgovind) [reply src/Microsoft.Data/StringColumn.cs]: Yes for the necessary question. This seemed more efficient to me than appending and doubling the size everytime? 
I don't see a more performant way to appending nulls to a list. If I had an array, I could've just initialized it. I could use LINQ, but I'm not so sure that that is more performant?

File: src/Microsoft.Data/BaseColumn.cs [hunk 26]
```diff
@@ -64,24 +67,26 @@ protected set
         /// <returns></returns>
         public virtual BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapIndices = false) => throw new NotImplementedException();
 
+        public virtual BaseColumn CloneAndAppendNulls(BaseColumn mapIndices = null, bool invertMapIndices = false) => throw new NotImplementedException();
```
Comment (eerhardt) [inline src/Microsoft.Data/BaseColumn.cs]: Can `appendNulls` just be an option to the existing `Clone` method?

**Comment by eerhardt** [comment]
> Oh also, how do folks reply to comments like this on GitHub? I don't see a "Reply" option, just an "Edit" option

Typically, you reply to comments like this on GitHub at the bottom, and using a `>` quote of the comment you are replying to.

File: src/Microsoft.Data/BaseColumn.cs (hunk 26 above)
Comment (pgovind) [reply src/Microsoft.Data/BaseColumn.cs]: It could. I'll put it in https://github.com/dotnet/corefxlab/pull/2687

**Comment by pgovind** [comment]
Thanks for the review @gfoidl  and @eerhardt
